dist/transport.go

```diff
@@ -0,0 +1,186 @@
+package dist
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/image"
+	"github.com/containers/image/v5/transports"
+	"github.com/containers/image/v5/types"
+	"github.com/pkg/errors"
+)
+
+//nolint (gochecknoinits)
+func init() {
+	transports.Register(Transport)
+}
+
+// nolint (gochecknoglobals)
+var Transport = distTransport{}
+
+type distTransport struct{}
+
+func (s distTransport) Name() string {
+	return "dist"
+}
+
+func splitReference(ref string) (fullname, server string, port int, err error) {
+	port = 8080
+	err = nil
+
+	if ref[0] == '/' {
+		fullname = ref[1:]
+		return fullname, "", port, nil
+	}
+
+	fields := strings.SplitN(ref, "/", 2)
+	subFields := strings.Split(fields[0], ":")
+
+	if len(subFields) > 2 {
+		err = fmt.Errorf("bad server:port")
+		return "", "", -1, err
+	}
+
+	server = subFields[0]
+
+	if len(subFields) == 2 {
+		port, err = strconv.Atoi(subFields[1])
+		if err != nil {
+			return "", "", -1, err
+		}
+
+		if port < 1 || port > 65535 {
+			err = fmt.Errorf("bad port %d", port)
+			return "", "", -1, err
+		}
+	}
+
+	fullname = fields[1]
+
+	return fullname, server, port, nil
+}
+
+// NOTE - the transport interface is defined in types/types.go.
+// Valid uris are:
+//    dist:///name1/name2/tag
+//    dist://server/name1/name2/name3/tag
+// The tag can be separated by either / or :
+//    dist://server:port/name1/name2/name3/tag
+//    dist://server:port/name1/name2/name3:tag
+// So the reference passed in here would be e.g.
+//    ///name1/name2/tag
+//    //server:port/name1/name2/tag
+func (s distTransport) ParseReference(reference string) (types.ImageReference, error) {
+	if !strings.HasPrefix(reference, "//") {
+		return nil, errors.Errorf("dist image reference %s does not start with //", reference)
+	}
+
+	fullname, server, port, err := splitReference(reference[2:])
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed parsing reference: '%s'", reference)
+	}
+
+	// support : for tag separateion
+	var name, tag string
+
+	fields := strings.Split(fullname, ":")
+
+	if len(fields) != 2 || len(fields[0]) == 0 || len(fields[1]) == 0 {
```

Comment:
So, `:tag` is mandatory, and the `/tag` syntax is invalid, after all?

No `:latest` default? (I could well live with this decision, but it’s an important and I imagine controversial one.)